### PR TITLE
Ignore empty list during JSON serialization

### PR DESCRIPTION
**Pull request by live1206**

### CODE REVIEW

File: docs/specs/moniker.yml [hunk 1]
```diff
@@ -117,7 +115,7 @@ outputs:
       "files": [
           { "url": "/docs/a", "path": "218c13d0/docs/a.json", "monikers": [ "netcore-1.1" ] },
           { "url": "/docs/a", "path": "4667fedf/docs/a.json", "monikers": [ "netcore-2.0" ] },
-          { "url": "/docs/a", "path": "docs/a.json", "monikers": [] },
+          { "url": "/docs/a", "path": "docs/a.json" },
```
Comment (yufeih) [inline docs/specs/moniker.yml]: `!monikers` to test moniker list is emty

File: docs/specs/git.yml [hunk 2]
```diff
@@ -194,8 +194,7 @@ outputs:
   docs/index.json: |
     {
       "author": 
-        { "name": "charlie", "profile_url": "https://github.com/charlie", "display_name": "Charlie", "id": "3" },
-      "contributors": []
+        { "name": "charlie", "profile_url": "https://github.com/charlie", "display_name": "Charlie", "id": "3" }
     }
```
Comment (yufeih) [inline docs/specs/git.yml]: `!contributors` to assert contributors are empty

File: src/docfx/lib/JsonUtility.cs [hunk 3]
```diff
@@ -482,6 +482,31 @@ private static void SetFieldWritable(MemberInfo member, JsonProperty prop)
...
+            protected override JsonProperty CreateProperty(MemberInfo member, MemberSerialization memberSerialization)
+            {
+                var prop = base.CreateProperty(member, memberSerialization);
+                var contract = DefaultSerializer.ContractResolver.ResolveContract(prop.PropertyType);
+                if (contract is JsonArrayContract)
+                {
+                    prop.ShouldSerialize =
+                          value =>
+                          {
+                              var obj = JToken.FromObject(value);
```
Comment (yufeih) [inline src/docfx/lib/JsonUtility.cs]: `JToken.FromObject` is to heavy to retrieve a single property.

File: src/docfx/lib/JsonUtility.cs [hunk 4]
```diff
@@ -482,6 +482,31 @@ private static void SetFieldWritable(MemberInfo member, JsonProperty prop)
...
+                                  return false;
+                              }
+                              return true;
+                          };
+                }
+                return prop;
+            }
+        }
+
         private sealed class JsonContractResolver : DefaultContractResolver
```
Comment (yufeih) [inline src/docfx/lib/JsonUtility.cs]: This this also derive from `SerializerContractResolver`?

Comment (live1206) [reply src/docfx/lib/JsonUtility.cs]: intended to split these two, should deserializer be camel case?

File: src/docfx/lib/JsonUtility.cs [hunk 5]
```diff
@@ -482,7 +482,32 @@ private static void SetFieldWritable(MemberInfo member, JsonProperty prop)
             }
         }
 
-        private sealed class JsonContractResolver : DefaultContractResolver
+        private class JsonContractResolver : CamelCasePropertyNamesContractResolver
+        {
+            protected override JsonProperty CreateProperty(MemberInfo member, MemberSerialization memberSerialization)
+            {
+                var prop = base.CreateProperty(member, memberSerialization);
+                var contract = DefaultSerializer.ContractResolver.ResolveContract(prop.PropertyType);
```
Comment (yufeih) [inline src/docfx/lib/JsonUtility.cs]: nit: consider extract this to a method like `don't serialize empty array`

File: docs/specs/moniker.yml [hunk 6]
```diff
@@ -100,7 +100,7 @@ outputs:
   docs/a.json: |
     {
       "content": "<p>No version</p>",
-      "monikers": []
+      "!monikers": null
```
Comment (OsmondJiang) [inline docs/specs/moniker.yml]: would you please help to add test for empty Dictionary?

Comment (live1206) [reply docs/specs/moniker.yml]: for json.net, Dictionary is JObject

**Comment by live1206** [comment]
Calling `DefaultSerializer.ContractResolver.ResolveContract` in `GetProperty` during de-serialization, will cause infinite recusion if the data contract is self referenced like the original `TestData`.

Since this logic is only needed for serialization, palling to move it to a dedicated contract resolver for serialization

**Comment by yufeih** [comment]
> Calling `DefaultSerializer.ContractResolver.ResolveContract` in `GetProperty` during de-serialization, will cause infinite recusion if the data contract is self referenced like the original `TestData`.
> 
> Since this logic is only needed for serialization, palling to move it to a dedicated contract resolver for serialization

Use a dedicated `DefaultContractResolver` to check if a type is array, that can avoid these recursion.

File: src/docfx/lib/JsonUtility.cs (hunk 5 above)
Comment (yufeih) [inline src/docfx/lib/JsonUtility.cs]: only do this if `ShouldSerialize` is null

Comment (live1206) [reply src/docfx/lib/JsonUtility.cs]: https://github.com/JamesNK/Newtonsoft.Json/blob/509643a8952ce731e0207710c429ad6e67dc43db/Src/Newtonsoft.Json/Serialization/CamelCasePropertyNamesContractResolver.cs#L69
There is a cache from type to JsonContract, the effort is minor
